src/theory/model_manager.cpp

```diff
@@ -162,5 +162,63 @@ bool ModelManager::collectModelBooleanVariables()
   return true;
 }
 
+void ModelManager::collectAssertedTerms(TheoryId tid,
+                                        std::set<Node>& termSet,
+                                        bool includeShared) const
+{
+  Theory* t = d_te.theoryOf(tid);
+  // Collect all terms appearing in assertions
+  context::CDList<Assertion>::const_iterator assert_it = t->facts_begin(),
+                                             assert_it_end = t->facts_end();
+  for (; assert_it != assert_it_end; ++assert_it)
+  {
+    collectTerms(tid, *assert_it, termSet);
+  }
+
+  if (includeShared)
+  {
+    // Add terms that are shared terms
+    context::CDList<TNode>::const_iterator shared_it = t->shared_terms_begin(),
+                                           shared_it_end =
+                                               t->shared_terms_end();
+    for (; shared_it != shared_it_end; ++shared_it)
+    {
+      collectTerms(tid, *shared_it, termSet);
+    }
+  }
+}
+
+void ModelManager::collectTerms(TheoryId tid,
+                                TNode n,
+                                std::set<Node>& termSet) const
+{
+  const std::set<Kind>& irrKinds = d_model->getIrrelevantKinds();
+  std::vector<TNode> visit;
+  TNode cur;
+  visit.push_back(n);
+  do
+  {
+    cur = visit.back();
+    visit.pop_back();
+    if (termSet.find(cur) != termSet.end())
+    {
+      // already visited
+      continue;
+    }
+    Kind k = cur.getKind();
+    // only add to term set if a relevant kind
+    if (irrKinds.find(k) == irrKinds.end())
```

Comment:
Shouldn't we also make sure that `Theory::theoryOf(cur) == tid`? Also, aren't we traversing irrelevant kinds multiple times?

Reply:
The former should always hold, I can add it as an assertion.

We are traversing terms with irrelevant kinds multiple times.  This isn't worth caching since all terms with irrelevant kinds are predicate symbols and cant be nested.  It's fairly unlikely that this will change in the future (I can't currently imagine any non-predicate symbols that are irrelevant).

Reply:
After discussing this offline, this is ok because the children of an irrelevant kind are cached if they are visited multiple times.